test/functional/test_short_deck.py

```diff
@@ -221,6 +231,148 @@ def test_call_action_sequence(n_players):
                 # Loop through the action history and make sure the bad
                 # sequence has not happened.
                 for i in range(len(no_fold_action_history)):
-                    history_slice = no_fold_action_history[i:i + len(bad_seq)]
+                    history_slice = no_fold_action_history[i : i + len(bad_seq)]
                     assert history_slice != bad_seq
             state = state.apply_action(random_action)
+
+
+@pytest.mark.parametrize("n_players", [2, 3])
+def test_action_sequence(n_players: int):
+    """
+    Check each round against validated action sequences to ensure the state class is
+    working correctly.
+    """
+    # Seed the random number generation so things are procedural/reproducable.
+    seed(42)
+    directory = "research/size_of_problem/action_sequences.pkl"
+    action_sequences = _load_action_sequences(directory)
+    for i in range(200):
+        state, _ = _new_game(n_players=n_players, small_blind=50, big_blind=100)
+
+        betting_stage_dict = {
+            "pre_flop": {"action_sequence": [], "n_players": 0},
+            "flop": {"action_sequence": [], "n_players": 0},
+            "turn": {"action_sequence": [], "n_players": 0},
+            "river": {"action_sequence": [], "n_players": 0},
+        }
+        betting_stage = None
+
+        while state.betting_stage not in {"show_down", "terminal"}:
+            if betting_stage != state.betting_stage:
+                betting_stage_dict[state.betting_stage][
+                    "n_players"
+                ] = state.n_players_started_round
+                betting_stage = state.betting_stage
+            uniform_probability: float = 1 / len(state.legal_actions)
+            probabilities = np.full(len(state.legal_actions), uniform_probability)
+            random_action: str = np.random.choice(state.legal_actions, p=probabilities)
+
+            betting_stage_dict[state.betting_stage]["action_sequence"].append(
+                random_action
+            )
+            state = state.apply_action(random_action)
+
+        for betting_stage in betting_stage_dict.keys():
+            if betting_stage_dict[betting_stage]["action_sequence"]:
+                n_players_started_round = betting_stage_dict[betting_stage]["n_players"]
+                action_sequence = betting_stage_dict[betting_stage]["action_sequence"]
+                possible_sequences = action_sequences[n_players_started_round]
+
+                assert action_sequence in possible_sequences
```

Comment:
I'm a little confused here, hoping you can help me here. Is this effectively a regression test? I.e you load previously computed histories into `action_sequences`, and then check that the current logic reproduces the same or similar sequences in `betting_stage_dict`?

Reply:
Yes, I don't know what constitutes a regression test. But action_sequences.py generates all possible action sequences using DFS of the game tree using the ShortDeckPokerState class for both 3 and 2 players. 

I then validated the actions with my eyeballs to make sure A) they were valid B) they are valid for all betting rounds given the number of players that started the round 

Having validated that I loaded them up to make sure the state class doesn't validate action sequence logic in the future.

Only one we may reconsider at a later point is the # of raises that are allowed.

Reply:
Should be noted that I believe all action sequences are the same for each round depending on the number of players that start the round.